hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/RowDataStreamWriteFunction.java

```diff
@@ -0,0 +1,492 @@
+    this.confirming = true;
+
+    writeMetrics.endDataFlush();
+    writeMetrics.resetAfterCommit();
+  }
+
+  private void registerMetrics() {
+    MetricGroup metrics = getRuntimeContext().getMetricGroup();
+    writeMetrics = new FlinkStreamWriteMetrics(metrics);
+    writeMetrics.registerMetrics();
+  }
+
+  protected List<WriteStatus> writeRecords(
+      String instant,
+      RowDataBucket rowDataBucket) {
+    writeMetrics.startFileFlush();
+
+    Iterator<BinaryRowData> rowItr =
+        new MutableIteratorWrapperIterator<>(
+            rowDataBucket.getDataIterator(), () -> new BinaryRowData(rowType.getFieldCount()));
+    Iterator<HoodieRecord> recordItr = new MappingIterator<>(
+        rowItr, rowData -> convertToRecord(rowData, rowDataBucket.getBucketInfo()));
+
+    List<WriteStatus> statuses = writeFunction.write(deduplicateRecordsIfNeeded(recordItr), rowDataBucket.getBucketInfo(), instant);
+    writeMetrics.endFileFlush();
+    writeMetrics.increaseNumOfFilesWritten();
+    return statuses;
+  }
+
+  protected HoodieFlinkRecord convertToRecord(RowData dataRow, BucketInfo bucketInfo) {
+    String key = keyGen.getRecordKey(dataRow);
+    Comparable<?> preCombineValue = preCombineFieldExtractor.getPreCombineField(dataRow);
+    HoodieOperation operation = HoodieOperation.fromValue(dataRow.getRowKind().toByteValue());
+    HoodieKey hoodieKey = new HoodieKey(key, bucketInfo.getPartitionPath());
+    return new HoodieFlinkRecord(hoodieKey, operation, preCombineValue, dataRow);
+  }
+
+  protected Iterator<HoodieRecord> deduplicateRecordsIfNeeded(Iterator<HoodieRecord> records) {
+    if (config.get(FlinkOptions.PRE_COMBINE)) {
+      // todo: sort by record key, lazy merge during iterating, default for COW.
+      return records;
+    } else {
+      return records;
+    }
+  }
+
+  private PreCombineFieldExtractor getPreCombineFieldExtractor(Configuration conf, RowType rowType) {
+    String preCombineField = OptionsResolver.getPreCombineField(conf);
+    if (StringUtils.isNullOrEmpty(preCombineField)) {
+      // return a dummy extractor.
+      return rowData -> HoodieRecord.DEFAULT_ORDERING_VALUE;
+    }
+    int preCombineFieldIdx = rowType.getFieldNames().indexOf(preCombineField);
+    LogicalType fieldType = rowType.getChildren().get(preCombineFieldIdx);
+    RowData.FieldGetter preCombineFieldGetter = RowData.createFieldGetter(fieldType, preCombineFieldIdx);
+
+    // currently the log reader for flink is avro reader, and it merges records based on ordering value
+    // in form of AVRO format, so here we align the data format with reader.
+    // TODO refactor this after RowData reader is supported, HUDI-9146.
+    RowDataToAvroConverters.RowDataToAvroConverter fieldConverter =
+        RowDataToAvroConverters.createConverter(fieldType, conf.get(FlinkOptions.WRITE_UTC_TIMEZONE));
+    Schema fieldSchema = AvroSchemaConverter.convertToSchema(fieldType, preCombineField);
```

Comment:
Not sure why the orderingValue type matters for write

Reply:
orderingValue will be stored in HoodieDeleteBlock for delete record, and during reading it will be used for merging, so we should dealing with the value type carefully. Now, we just follow convention as before.